### PR TITLE
Add stronger alternatives to `align_to`

### PR DESCRIPTION
**Pull request by oli-obk**
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/105296)*

r? @ghost

some code for the [discussion](https://rust-lang.zulipchat.com/#narrow/stream/146212-t-compiler.2Fconst-eval/topic/align_to.20and.20other.20APIs.20with.20surprising.20const.20behavior)

ACP: https://github.com/rust-lang/libs-team/issues/164

### CODE REVIEW

File: library/core/src/slice/mod.rs [hunk 1]
```diff
@@ -3594,6 +3602,166 @@ impl<T> [T] {
         }
     }
 
+    /// Get a subslice where the first element and the element one past the end is aligned
```
Comment (RalfJung) [inline library/core/src/slice/mod.rs:3605]: This is a bit confusing as it indicates that there might be an element after the end.

I think it'd be better to talk about the first element being aligned and the size of the return slice being a multiple of the alignment.

File: library/core/src/slice/mod.rs [hunk 2]
```diff
@@ -3594,6 +3602,166 @@ impl<T> [T] {
...
+    /// the slice element's alignment.
+    ///
+    /// # Examples
+    ///
+    /// Basic usage:
+    ///
+    /// ```
+    /// unsafe {
+    ///     let mut bytes: [u8; 7] = [1, 2, 3, 4, 5, 6, 7];
+    ///     let ints = bytes.aligned_subslice_mut::<u32>();
```
Comment (RalfJung) [inline library/core/src/slice/mod.rs:3659]: This doesn't match the signature? It matches what I kind of expected though. But then the function can't be safe any more. Which seems fine?

File: library/core/src/slice/mod.rs [hunk 3]
```diff
@@ -3594,6 +3602,166 @@ impl<T> [T] {
...
+    ///     let smaller_ints = bytes.transmute_elements::<u16>();
+    ///     assert_eq!(smaller_ints.len(), 4);
+    /// }
+    /// ```
+    #[must_use]
+    #[unstable(feature = "slice_align_to_ish", issue = "none")]
+    pub const unsafe fn transmute_elements<U>(&self) -> &[U] {
+        const {
+            let u = crate::mem::size_of::<U>();
+            let t = crate::mem::size_of::<T>();
```
Comment (RalfJung) [inline library/core/src/slice/mod.rs:3708]: These non-descriptive variable names make the code kind of hard to read.

File: library/core/src/slice/mod.rs [hunk 4]
```diff
@@ -3602,8 +3602,8 @@ impl<T> [T] {
         }
     }
 
-    /// Get a subslice where the first element and the element one past the end is aligned
-    /// to a given alignment.
+    /// Get a subslice where the first element is aligned to a given alignment and the length
+    /// of the slice is a multiple of the alignment.
```
Comment (RalfJung) [inline library/core/src/slice/mod.rs:3606]: The size, not the length -- right?

File: library/core/src/slice/mod.rs [hunk 5]
```diff
@@ -3594,6 +3602,166 @@ impl<T> [T] {
...
+            let align_u = crate::mem::align_of::<U>();
+            let align_t = crate::mem::align_of::<T>();
+            assert!(align_u <= align_t);
+        };
+        let align_u = crate::mem::size_of::<U>();
+        let align_t = crate::mem::size_of::<T>();
+        // SAFETY: The size of the slice is such that with the new element size, all new
+        // elements are still within the bounds of the original slice. The change in element
+        // type is something the caller needs to make sure is sound.
+        unsafe { from_raw_parts(self.as_ptr() as *const _, self.len() * align_t / align_u) }
```
Comment (tmiasko) [inline library/core/src/slice/mod.rs:3721]: ` self.len() * size_t / size_u`

File: library/core/src/slice/mod.rs [hunk 6]
```diff
@@ -3594,6 +3602,166 @@ impl<T> [T] {
...
+            "aligned_subslice only works for alignments that are multiples of the element's size"
+        );
+        let offset = self.as_ptr().addr() % align;
+        // SAFETY: See the `align_to_mut` method for the detailed safety comment.
+        let end_offset = unsafe { self.as_ptr().offset(self.len() as isize) }.addr() % align;
+        let end = self.len() - (align / size - end_offset);
+        &mut self[offset..end]
+    }
+
+    /// Transmute the slice elements to another (smaller) type, giving a longer slice.
```
Comment (Mark-Simulacrum) [inline library/core/src/slice/mod.rs:3683]: Is there a reason to identify it as *smaller*? This seems useful for 1:1 mapping, too, e.g. `u8` <> `i8`.

File: library/core/src/slice/mod.rs [hunk 7]
```diff
@@ -3594,6 +3602,166 @@ impl<T> [T] {
...
+    ///
+    /// The returned slice will have multiple elements per element of the original slice.
+    ///
+    /// # Safety
+    ///
+    /// This method is essentially a `transmute` with respect to the elements in the returned
+    /// middle slice, so all the usual caveats pertaining to `transmute::<T, U>` also apply here.
```
Comment (Mark-Simulacrum) [inline library/core/src/slice/mod.rs:3690]: "middle slice" looks leftover here.

**Comment by Lokathor** [comment]
One fn seems to change to bigger elements and the other to smaller elements. I'm not clear why it's two separate functions, and also the naming doesn't really suggest which is which.

**Comment by oli-obk** [comment]
> One fn seems to change to bigger elements and the other to smaller elements. I'm not clear why it's two separate functions, and also the naming doesn't really suggest which is which.

the `transmute*` methods are now for all kinds of size changes and document where they won't work (if the alignment gets bigger) and suggest to use the other method instead.

File: library/core/src/slice/mod.rs [hunk 8]
```diff
@@ -3473,6 +3473,10 @@ impl<T> [T] {
     /// should depend on that, not its correctness. It is permissible for all of the input data to
     /// be returned as the prefix or suffix slice.
     ///
+    /// If this behaviour is not what you desire, as you don't need fallback paths for the bytes
```
Comment (RalfJung) [inline library/core/src/slice/mod.rs:3476]: "need" should probably be "want" or "have"?

**Comment by jswrenn** [comment]
> Minor naming thought: Would it be worth leaving space for a future safe-transmute version to have the "ideal" name, and have the unsafe versions be `_unchecked` or similar?
> 
> (Perhaps `transmute` is enough to communicate the unsafety, though, and we'll use a different word for the safe versions.)

FWIW, I'm leaning towards suggesting that the safe transmute APIs use the word "reinterpret" instead of "transmute".

**Comment by scottmcm** [comment]
@jswrenn Sounds great -- [I've long been a fan](https://github.com/rust-lang/rust/issues/49792#issuecomment-379638786) of that choice :)

**Comment by oli-obk** [comment]
r? libs-api

File: library/core/src/slice/mod.rs [hunk 9]
```diff
@@ -3607,6 +3615,200 @@ impl<T> [T] {
...
+    /// ```
+    /// #![feature(slice_align_to_ish)]
+    /// let bytes: [u8; 7] = [1, 2, 3, 4, 5, 6, 7];
+    /// let ints = bytes.aligned_subslice(std::mem::align_of::<u32>());
+    /// assert_eq!(ints.len(), 1);
+    /// ```
+    #[must_use]
+    #[unstable(feature = "slice_align_to_ish", issue = "none")]
+    #[inline]
+    pub fn aligned_subslice(&self, align: usize) -> &[T] {
```
Comment (the8472) [inline library/core/src/slice/mod.rs:3639]: Could the `align` be const? Does one ever need dynamic alignment?

File: library/core/src/slice/mod.rs [hunk 10]
```diff
@@ -3607,6 +3615,200 @@ impl<T> [T] {
...
+            assert!(
+                self.len() * size_u % size_t == 0,
+                "input slice does not fit exactly into a slice of the output element type"
+            );
+        } else {
+            assert!(
+                self.len() * size_t % size_u == 0,
+                "input slice does not fit exactly into a slice of the output element type"
+            );
+        }
```
Comment (the8472) [inline library/core/src/slice/mod.rs:3746]: Do the panics get eliminated when the types are exact multiples in size and alignment?

**Comment by anden3** [comment]
Hello @oli-obk! I just want to ping you as part of the triage procedure to let you know that this PR has received reviews :)

**Comment by oli-obk** [comment]
The PR is blocked on https://github.com/rust-lang/libs-team/issues/164

**Pull request closed without merging**